// @vivliostyle/cli@4.3.2/dist/commands/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const fs_1 = __importDefault(require("fs"));
const upath_1 = __importDefault(require("upath"));
const container_1 = require("../container");
const util_1 = require("../util");
const init_parser_1 = require("./init.parser");
try {
    const program = init_parser_1.setupInitParserProgram();
    program.parse(process.argv);
    const options = program.opts();
    init({
        title: options.title,
        author: options.author,
        language: options.language,
        size: options.size,
        theme: options.theme,
    }).catch(util_1.gracefulError);
}
catch (err) {
    util_1.gracefulError(err);
}
async function init(cliFlags) {
    const vivliostyleConfigPath = upath_1.default.join(util_1.cwd, 'vivliostyle.config.js');
    if (fs_1.default.existsSync(vivliostyleConfigPath)) {
        return util_1.log(`${chalk_1.default.yellow('vivliostyle.config.js already exists. aborting.')}`);
    }
    // prettier-ignore
    const vivliostyleConfig = `module.exports = {
  title: '${cliFlags.title || 'Principia'}', // populated into 'publication.json', default to 'title' of the first entry or 'name' in 'package.json'.
  author: '${cliFlags.author || 'Isaac Newton'}', // default to 'author' in 'package.json' or undefined
  ${cliFlags.language ? '' : '// '}language: '${cliFlags.language || 'la'}',
  ${cliFlags.size ? '' : '// '}size: '${cliFlags.size || 'A4'}',
  ${cliFlags.theme ? '' : '// '}theme: '${cliFlags.theme || ''}', // .css or local dir or npm package. default to undefined
  image: '${container_1.CONTAINER_IMAGE}',
  entry: [ // **required field**
    // 'introduction.md', // 'title' is automatically guessed from the file (frontmatter > first heading)
    // {
    //   path: 'epigraph.md',
    //   title: 'おわりに', // title can be overwritten (entry > file),
    //   theme: '@vivliostyle/theme-whatever' // theme can be set indivisually. default to root 'theme'
    // },
    // 'glossary.html' // html is also acceptable
  ], // 'entry' can be 'string' or 'object' if there's only single markdown file
  // entryContext: './manuscripts', // default to '.' (relative to 'vivliostyle.config.js')
  // output: [ // path to generate draft file(s). default to '{title}.pdf'
  //   './output.pdf', // the output format will be inferred from the name.
  //   {
  //     path: './book',
  //     format: 'webpub',
  //   },
  // ],
  // workspaceDir: '.vivliostyle', // directory which is saved intermediate files.
  // toc: true, // whether generate and include ToC HTML or not, default to 'false'.
  // cover: './cover.png', // cover image. default to undefined.
  // vfm: { // options of VFM processor
  //   hardLineBreaks: true, // converts line breaks of VFM to <br> tags. default to 'false'.
  //   disableFormatHtml: true, // disables HTML formatting. default to 'false'.
  // },
};
`;
    fs_1.default.writeFileSync(vivliostyleConfigPath, vivliostyleConfig);
    util_1.log(`Successfully created ${chalk_1.default.cyan('vivliostyle.config.js')}`);
}
exports.default = init;
